pkg/reader/ccip.go

```diff
@@ -439,29 +441,100 @@ func (r *ccipChainReader) Nonces(
 	return res, nil
 }
 
-func (r *ccipChainReader) GasPrices(ctx context.Context, chains []cciptypes.ChainSelector) ([]cciptypes.BigInt, error) {
-	if err := validateWriterExistence(r.contractWriters, chains...); err != nil {
-		return nil, err
+func (r *ccipChainReader) GetAvailableChainsFeeComponents(
+	ctx context.Context,
+) map[cciptypes.ChainSelector]types.ChainFeeComponents {
+	feeComponents := make(map[cciptypes.ChainSelector]types.ChainFeeComponents, len(r.contractWriters))
+	for chain, chainWriter := range r.contractWriters {
+		feeComponent, err := chainWriter.GetFeeComponents(ctx)
+		if err != nil {
+			r.lggr.Errorw("failed to get chain fee components for chain %d: %w", chain, err)
+			continue
+		}
+		feeComponents[chain] = *feeComponent
 	}
+	return feeComponents
+}
 
-	eg := new(errgroup.Group)
-	gasPrices := make([]cciptypes.BigInt, len(chains))
-	for i, chain := range chains {
-		i, chain := i, chain
-		eg.Go(func() error {
-			gasPrice, err := r.contractWriters[chain].GetFeeComponents(ctx)
-			if err != nil {
-				return fmt.Errorf("failed to get gas price: %w", err)
-			}
-			gasPrices[i] = cciptypes.NewBigInt(gasPrice.ExecutionFee)
-			return nil
-		})
+func (r *ccipChainReader) GetWrappedNativeTokenPriceUSD(
+	ctx context.Context,
+	selectors []cciptypes.ChainSelector,
+) map[cciptypes.ChainSelector]cciptypes.BigInt {
+	// 1. Call chain's router to get native token address https://github.com/smartcontractkit/chainlink/blob/60e8b1181dd74b66903cf5b9a8427557b85357ec/contracts/src/v0.8/ccip/Router.sol#L189:L191
+	// nolint:lll
+	// 2. Call FeeQuoter to get native tokens price  https://github.com/smartcontractkit/chainlink/blob/60e8b1181dd74b66903cf5b9a8427557b85357ec/contracts/src/v0.8/ccip/FeeQuoter.sol#L229-L229
+	prices := make(map[cciptypes.ChainSelector]cciptypes.BigInt, len(selectors))
+	for _, chain := range selectors {
+		reader, ok := r.contractReaders[chain]
+		if !ok {
+			r.lggr.Warnw("contract reader not found", "chain", chain)
+			continue
+		}
+
+		var nativeTokenAddress types2.Account
+		err := reader.ExtendedGetLatestValue(
+			ctx,
+			consts.ContractNameRouter,
+			consts.MethodNameRouterGetWrappedNative,
+			primitives.Unconfirmed,
+			nil,
+			&nativeTokenAddress)
+
+		if err != nil {
+			r.lggr.Errorw("failed to get native token address", "chain", chain, "err", err)
+			continue
+		}
+
+		var price *big.Int
+		err = reader.ExtendedGetLatestValue(
+			ctx,
+			consts.ContractNameFeeQuoter,
+			consts.MethodNameFeeQuoterGetTokenPrices,
+			primitives.Unconfirmed,
+			map[string]any{
+				"token": nativeTokenAddress,
+			},
+			&price,
+		)
+		if err != nil {
+			r.lggr.Errorw("failed to get native token price", "chain", chain, "err", err)
+			continue
+		}
+		prices[chain] = cciptypes.NewBigInt(price)
```

Comment:
We have that concurrent call pattern used in other methods, but I'm not sure it's worth the extra complexity. Maybe we keep it simple like this and optimize with the Batch calls later.

Reply:
Added TODO 👍